crates/turbopack-test-utils/src/snapshot.rs

```diff
@@ -0,0 +1,178 @@
+use std::{
+    collections::{HashMap, HashSet},
+    env, fs,
+};
+
+use anyhow::{anyhow, bail, Context, Result};
+use once_cell::sync::Lazy;
+use similar::TextDiff;
+use turbo_tasks::{TryJoinIterExt, ValueToString};
+use turbo_tasks_fs::{
+    DirectoryContent, DirectoryEntry, DiskFileSystemVc, File, FileContent, FileSystemEntryType,
+    FileSystemPathVc,
+};
+use turbo_tasks_hash::encode_hex;
+use turbopack_core::{
+    asset::{AssetContent, AssetContentVc},
+    issue::PlainIssueReadRef,
+};
+
+// Updates the existing snapshot outputs with the actual outputs of this run.
+// e.g. `UPDATE=1 cargo test -p turbopack-tests -- test_my_pattern`
+static UPDATE: Lazy<bool> = Lazy::new(|| env::var("UPDATE").unwrap_or_default() == "1");
+
+pub async fn snapshot_issues<I: IntoIterator<Item = PlainIssueReadRef>>(
+    captured_issues: I,
+    issues_path: FileSystemPathVc,
+    workspace_root: &str,
+) -> Result<()> {
+    let expected_issues = expected(issues_path).await?;
+    let mut seen = HashSet::new();
+    for plain_issue in captured_issues.into_iter() {
+        let hash = encode_hex(plain_issue.internal_hash());
+
+        // We replace "*" because it's not allowed for filename on Windows.
+        let path = issues_path.join(&format!(
+            "{}-{}.txt",
+            plain_issue.title.replace('*', "__star__"),
+            &hash[0..6]
+        ));
+        seen.insert(path);
+
+        // Annoyingly, the PlainIssue.source -> PlainIssueSource.asset ->
+        // PlainAsset.path -> FileSystemPath.fs -> DiskFileSystem.root changes
+        // for everyone.
+        let content = format!("{:#?}", plain_issue).replace(workspace_root, "WORKSPACE_ROOT");
```

Comment:
This is still not resolved. We can't use the `Debug` trait of the read ref, because it debugs the Cells, not the Values. We need to receive a `Vc` here, call `vc.dbg().await`, and diff that.